vm/src/stdlib/signal.rs

```diff
@@ -23,51 +23,64 @@ const SIG_IGN: libc::sighandler_t = 1;
 
 extern "C" fn run_signal(signum: i32) {
-    unsafe {
-        TRIGGERS[signum as usize].store(true, Ordering::Relaxed);
-    }
+    TRIGGERS[signum as usize].store(true, Ordering::Relaxed);
 }
 
-fn signal(
-    signalnum: PyIntRef,
-    handler: PyObjectRef,
-    vm: &VirtualMachine,
-) -> PyResult<Option<PyObjectRef>> {
-    if !vm.isinstance(&handler, &vm.ctx.function_type())?
-        && !vm.isinstance(&handler, &vm.ctx.bound_method_type())?
-        && !vm.isinstance(&handler, &vm.ctx.builtin_function_or_method_type())?
-    {
-        return Err(vm.new_type_error("Hanlder must be callable".to_string()));
-    }
-    let signal_module = vm.import("signal", &vm.ctx.new_tuple(vec![]), 0)?;
-    let sig_dfl = vm.get_attribute(signal_module.clone(), "SIG_DFL")?;
-    let sig_ign = vm.get_attribute(signal_module, "SIG_IGN")?;
-    let signalnum = signalnum.as_bigint().to_i32().unwrap();
-    check_signals(vm);
-    let sig_handler = if handler.is(&sig_dfl) {
-        SIG_DFL
-    } else if handler.is(&sig_ign) {
-        SIG_IGN
+fn assert_in_range(signum: i32, vm: &VirtualMachine) -> PyResult<()> {
+    if (1..=NSIG as i32).contains(&signum) {
+        Ok(())
     } else {
-        run_signal as libc::sighandler_t
+        Err(vm.new_value_error("signal number out of range".to_owned()))
+    }
+}
+
+fn signal(signalnum: i32, handler: PyObjectRef, vm: &VirtualMachine) -> PyResult<PyObjectRef> {
+    assert_in_range(signalnum, vm)?;
+
+    let sig_handler = match usize::try_from_object(vm, handler.clone()).ok() {
+        Some(SIG_DFL) => SIG_DFL,
+        Some(SIG_IGN) => SIG_IGN,
+        None if vm.is_callable(&handler) => run_signal as libc::sighandler_t,
+        _ => {
+            return Err(vm.new_type_error(
+                "signal handler must be signal.SIG_IGN, signal.SIG_DFL, or a callable object"
+                    .to_owned(),
+            ))
+        }
     };
+    check_signals(vm)?;
+
     let old = unsafe { libc::signal(signalnum, sig_handler) };
     if old == SIG_ERR {
         return Err(vm.new_os_error("Failed to set signal".to_string()));
     }
-    let old_handler = vm.signal_handlers.borrow_mut().insert(signalnum, handler);
+    #[cfg(unix)]
+    {
+        extern "C" {
+            fn siginterrupt(sig: i32, flag: i32);
+        }
+        unsafe {
+            siginterrupt(signalnum, 1);
+        }
+    }
+
+    let mut old_handler = handler;
+    std::mem::swap(
+        &mut vm.signal_handlers.borrow_mut()[signalnum as usize],
+        &mut old_handler,
+    );
     Ok(old_handler)
 }
 
-fn getsignal(signalnum: PyIntRef, vm: &VirtualMachine) -> PyResult<Option<PyObjectRef>> {
-    let signalnum = signalnum.as_bigint().to_i32().unwrap();
-    Ok(vm.signal_handlers.borrow_mut().get(&signalnum).cloned())
+fn getsignal(signalnum: i32, vm: &VirtualMachine) -> PyResult<PyObjectRef> {
+    assert_in_range(signalnum, vm)?;
+    Ok(vm.signal_handlers.borrow()[signalnum as usize].clone())
```

Comment:
We should probably check for signalnum boundaries in the array.

Reply:
That's what `assert_in_range` does.

Reply:
That's what `assert_in_range` does.

Reply:
Ah, I missed that. Thanks!